Frameworks/Foundation/NSDecimal.mm

```diff
@@ -0,0 +1,605 @@
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+#import <Starboard.h>
+#import <StubReturn.h>
+#import <Foundation/NSDecimal.h>
+#import <algorithm>
+
+#define IS_NEGATIVE(decimal) (decimal)->_isNegative
+
+#define IS_POSITIVE(decimal) !IS_NEGATIVE(decimal)
+
+#define IS_ZERO(decimal) (((decimal)->_length == 0) && IS_POSITIVE(decimal))
+
+#define INVALIDATE_DECIMAL(decimal) \
+    \
+decimal->_length = 0;               \
+    \
+decimal->_isNegative = YES;
+
+#define SET_TO_ZERO(decimal) \
+    \
+decimal->_length = 0;        \
+    \
+decimal->_isNegative = NO;   \
+    \
+NSDecimalCompact(decimal);
+
+#define VALIDATE_ARITHMETIC_INPUT(result, leftOperand, rightOperand)                 \
+    if (NSDecimalIsNotANumber(leftOperand) || NSDecimalIsNotANumber(rightOperand)) { \
+        INVALIDATE_DECIMAL(result);                                                  \
+        return NSCalculationOverflow;                                                \
+    }
+
+const int _MAX_EXP = 127;
+
+typedef struct {
+    NSDecimal decimal;
+    unsigned short* _mantissa;
+    unsigned short _remainder;
+    // only used for cases where we have carryover
+    unsigned int _extendedLen;
+} NSDecimalInternal;
+
+/**
+ @Status Interoperable
+*/
+void NSDecimalCopy(NSDecimal* destination, const NSDecimal* source) {
+    if (!source || !destination) {
+        return;
+    }
+
+    destination->_isNegative = source->_isNegative;
+    destination->_exponent = source->_exponent;
+    destination->_length = source->_length;
+    destination->_isCompact = source->_isCompact;
+    destination->_reserved = source->_reserved;
+    memcpy(&destination->_mantissa, &source->_mantissa, NSDecimalMaxSize * sizeof(unsigned short));
+}
+
+/**
+ @Status Interoperable
+*/
+BOOL NSDecimalIsNotANumber(const NSDecimal* dcm) {
+    THROW_NS_IF_FALSE(E_INVALIDARG, dcm != NULL);
+    return !dcm->_length && IS_NEGATIVE(dcm);
+}
+
+static void _InternalDecimalCopy(NSDecimal* destination, NSDecimalInternal* source) {
+    if (!source || !destination) {
+        return;
+    }
+    NSDecimalCopy(destination, &(source->decimal));
+    memcpy(destination->_mantissa, source->_mantissa, NSDecimalMaxSize * sizeof(unsigned short));
+}
+
+static NSCalculationError _subtraction(NSDecimal* result, const NSDecimal* left, const NSDecimal* right) {
+    // Note: All inputs have been normalized via NSDecimalNormalize
+    unsigned int commonLength = std::min(left->_length, right->_length);
+
+    unsigned int overflow = 1;
+
+    // subtract the numbers in the same size
+    for (int index = 0; index < commonLength; index++) {
+        overflow += USHRT_MAX + ((left->_mantissa)[index] - (right->_mantissa)[index]);
+        (result->_mantissa)[index] = overflow;
```

Comment:
note the mantissa is a 8byte, each byte represent a factor of USHRT_MAX + 1. when you down cast from int to mantissa, you are storing the (overflow & 0x0000FFFF)

---
In reply to: [80598557](https://github.com/Microsoft/WinObjC/pull/1032#discussion_r80598557) [](ancestors = 80598557)